### ISSUE TITLE
Accelerate Model Operations with GPU Support

### ISSUE THREAD

**Issue by smellycloud** [views-platform/views-stepshifter#31]
## **Title**
Accelerate Model Operations with GPU Support for `_predict_by_step` and Overall Workflow

## **Description**
The `_predict_by_step` method in the `StepshifterModel` and `HurdleModel` currently takes over 5 hours for PGM (priogrid-month) runs on fimbulthul. This feature request proposes migrating all model operations to GPU (CUDA/MPS) to drastically reduce computation time. The focus is on enabling GPU support for Darts-based models (e.g., `XGBModel`,` LightGBMModel`) and ensuring end-to-end GPU utilization during training and prediction.

## **Motivation**
* **Performance**: GPU acceleration can reduce runtime from hours to minutes, enabling faster iteration and scalability.
* **Modern Hardware Utilization**: Many systems now have GPUs (NVIDIA/CUDA or Apple MPS), and leveraging them maximizes resource efficiency.
* **Competitive Advantage**: GPU support is critical for handling larger datasets and complex models in production.

### 1. Enable GPU-Accelerated Model Training/Prediction
* **Model Configuration**:
**XGBoost**: Set `tree_method="gpu_hist"` and `device="cuda"` in XGBModel parameters.
**LightGBM**: Enable `device="gpu"` for `LightGBMModel`.
**Other Models**: Investigate GPU support for `LinearRegressionModel` (via PyTorch backend) and `RandomForest` (via cuML).

* **Automatic GPU Detection**: Add utility functions to detect CUDA/MPS availability and configure models accordingly:
```python
def get_device_params():
    if torch.cuda.is_available():
        return {"device": "cuda"}
    elif torch.backends.mps.is_available():
        return {"device": "mps"}
    else:
        return {}
````
**Parameter Injection**: Modify `_resolve_estimator` to inject GPU parameters dynamically during model initialization.

### 2. Data Pipeline Optimization
* Ensure input data (TimeSeries) is in GPU-compatible formats (e.g., float32 tensors).
* Use Darts’ `TorchForecastingModel` subclasses where possible to leverage PyTorch GPU acceleration.

### 3. Batch Prediction on GPU
* Refactor `_predict_by_step` to batch predictions across steps/sequences, minimizing CPU-GPU data transfers.

### 4. Error Handling and Fallback
* Gracefully fall back to CPU with warnings if GPU setup fails.
* Validate GPU memory usage to avoid OOM errors.

### Code Modifications
Example for XGBModel GPU Support:
```python
# In _resolve_estimator, inject GPU params conditionally
case "XGBModel":
    from darts.models import XGBModel
    params = {"tree_method": "gpu_hist", "device": "cuda"} if cuda_available else {}
    return partial(XGBModel, **params)
```

## **Alternatives Considered**
* **cuML Replacements**: Replacing scikit-learn models with RAPIDS/cuML equivalents was considered but rejected due to Darts integration complexity.
* **PyTorch Lightning**: Rewriting models with PyTorch Lightning for GPU orchestration was deemed too invasive for the current codebase.
* **Dask Parallelization**: Distributed CPU computation was explored but offers less speedup compared to GPU acceleration.

## **Additional Context**
**Dependencies**: Requires CUDA-toolkit or MPS-compatible PyTorch installations. Explore handling this in the TOML file.
**Benchmarking**: Do a run time comparison between GPU vs CPU to quantify the potential performance gains.

## **Checklist**
- [x] I have searched for existing feature requests that might be related to this one.
- [x] I have provided a detailed description of the feature request.
- [x] I have included all necessary information about the motivation and proposed solution.
- [x] I have attached relevant diagrams, mockups, or code snippets.

## **Actionable Tasks**
- [ ] Implement GPU detection and parameter injection logic.
- [ ] Update model configurations for XGBoost, LightGBM, and other GPU-compatible models.
- [ ] Refactor `_predict_by_step` to batch predictions on GPU.
- [ ] Add GPU memory management and error handling.
- [ ] Update documentation for GPU usage.

## **Error Handling & Logging**
- Log GPU configuration status (CUDA/MPS/CPU) at runtime.
- Catch GPU-specific exceptions (e.g., CUDA OOM) and fall back to CPU with warnings.

## **Future Flexibility**
- Abstract GPU/device management to simplify adding new accelerators (e.g., AMD ROCm).

## **Branch**
- Work on a new branch named `feature_gpu_acceleration`.
